src/ledger/database/lmdb.zig

```diff
@@ -0,0 +1,716 @@
+const std = @import("std");
+const c = @import("lmdb");
+const sig = @import("../../sig.zig");
+const database = @import("lib.zig");
+
+const Allocator = std.mem.Allocator;
+
+const BytesRef = database.interface.BytesRef;
+const ColumnFamily = database.interface.ColumnFamily;
+const IteratorDirection = database.interface.IteratorDirection;
+const Logger = sig.trace.Logger;
+
+const key_serializer = database.interface.key_serializer;
+const value_serializer = database.interface.value_serializer;
+
+pub fn LMDB(comptime column_families: []const ColumnFamily) type {
+    return struct {
+        allocator: Allocator,
+        env: *c.MDB_env,
+        dbis: []const c.MDB_dbi,
+        path: []const u8,
+
+        const Self = @This();
+
+        pub fn open(allocator: Allocator, logger: Logger, path: []const u8) anyerror!Self {
+            logger.info().log("Initializing LMDB");
+            const owned_path = try allocator.dupe(u8, path);
+
+            // create and open the database
+            const env = try ret(c.mdb_env_create, .{});
+            try result(c.mdb_env_set_maxdbs(env, column_families.len));
+            try result(c.mdb_env_open(env, @ptrCast(path), 0, 0o700));
+
+            // begin transaction to create column families aka "databases" in lmdb
+            const txn = try ret(c.mdb_txn_begin, .{ env, null, 0 });
+            errdefer c.mdb_txn_abort(txn);
+
+            // allocate cf handles
+            const dbis = try allocator.alloc(c.MDB_dbi, column_families.len);
+            errdefer allocator.free(dbis);
+
+            // save cf handles
+            inline for (column_families, 0..) |cf, i| {
+                // open cf/database, creating if necessary
+                dbis[i] = try ret(c.mdb_dbi_open, .{
+                    txn,
+                    @as([*c]const u8, @ptrCast(cf.name)),
```

Comment:
This is not a translated file, it should never use C pointers.

Reply:
890d797380ee60f0a6513a5708ec0d79754fa586

Reply:
Two things here.
1. I took a quick glance at lmbd, and it calls `strlen` on the name passed in. This should be a `[*:0]const u8` type. Either call `allocator.dupeZ` on the name, or just store `cf.name` as a `[:0]const u8`.
2. There is no need for a `@ptrCast` here, doing `slice.ptr` yields `[*]T`.

Reply:
ac1e07da776c82a5ca62f1bae7590c8342d4aa1a
b8adf694e731c1828150f8c802193b60abcd45f7